tree/ntuple/v7/inc/ROOT/RDaos.hxx

```diff
@@ -0,0 +1,219 @@
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RDaos
+#define ROOT7_RDaos
+
+#include <ROOT/RStringView.hxx>
+#include <ROOT/TypeTraits.hxx>
+
+#include <daos.h>
+// Avoid depending on `gurt/common.h` as the only required declaration is `d_rank_list_free()`.
+// Also, this header file is known to provide macros that conflict with std::min()/std::max().
+extern "C" void d_rank_list_free(d_rank_list_t *rank_list);
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace ROOT {
+
+namespace Experimental {
+namespace Detail {
+class RDaosContainer;
+
+/**
+  \class RDaosPool
+  \brief A RDaosPool provides access to containers in a specific DAOS pool.
+  */
+class RDaosPool {
+   friend class RDaosContainer;
+private:
+   daos_handle_t fPoolHandle{};
+   daos_pool_info_t fPoolInfo{};
+   uuid_t fPoolUuid{};
+
+public:
+   RDaosPool() = delete;
+   RDaosPool(std::string_view poolUuid, std::string_view serviceReplicas);
+   ~RDaosPool();
+};
+
+/**
+  \class RDaosObject
+  \brief Provides low-level access to DAOS objects in a container.
+  */
+template <typename DKeyT, typename AKeyT>
+class RDaosObject {
+private:
+   daos_handle_t fObjectHandle;
+
+   // Provide support for UINT64 dkey/akey.
+   template <typename T, typename std::enable_if<!ROOT::TypeTraits::HasDataAndSize<T>::value, int>::type = 0>
+   static size_t key_size(T& x) { return sizeof(x); }
+
+   template <typename T, typename std::enable_if<!ROOT::TypeTraits::HasDataAndSize<T>::value, int>::type = 0>
+   static typename std::add_pointer<T>::type key_data(T& x) { return &x; }
+
+   // Provide support for std::string/std::vector dkey/akey.
+   template <typename T, typename std::enable_if<ROOT::TypeTraits::HasDataAndSize<T>::value, int>::type = 0>
+   static size_t key_size(T& x) { return x.size(); }
+
+   template <typename T, typename std::enable_if<ROOT::TypeTraits::HasDataAndSize<T>::value, int>::type = 0>
+   static typename T::pointer key_data(T& x) { return const_cast<typename T::pointer>(x.data()); }
+
+public:
+   using DistributionKey_t = DKeyT;
+   using AttributeKey_t = AKeyT;
+  
+   /// \brief Contains required information for a single fetch/update operation.
+   struct FetchUpdateArgs {
+      FetchUpdateArgs() = default;
+      FetchUpdateArgs(const FetchUpdateArgs&) = delete;
+      FetchUpdateArgs(FetchUpdateArgs&& fua);
```

Comment:
It is used by `std::vector<...>`, if the internal storage needs to be reallocated.  Even if we are calling `std::vector<...>::reserve()` (see the `RDaosContainer::VectorReadWrite()` function), it is not bad to keep it, as it might avoid problems in the future.